{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "fd5bfb8e-2dec-4d2e-b09f-329ea7f1da3e",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pyabraom import *\n",
    "from pynoma import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c22347df-bfba-481c-b228-fc3799d0eb5a",
   "metadata": {},
   "outputs": [],
   "source": [
    "rs = RegionSearch(3, 17, 43045690, 43045710)\n",
    "df, clinical_df = rs.get_data()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "35e917c1-1123-4e3d-89f3-a1dd792c073b",
   "metadata": {},
   "outputs": [],
   "source": [
    "reg = 89740838\n",
    "df = Search_region('hg38', 16, reg-50, reg+50)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "8fc17d3e-94b4-46b4-b449-bc10f6489bdf",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "rs = GeneSearch(3, 'BAK1')\n",
    "mad_df, clinical_df = rs.get_data()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "c41edd22-40b4-425c-9014-206bb2fbd42e",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "df = Search_gene('hg38', 'BAK1')\n",
    "dels = df.loc[df['Alternative'] == \"-\"].Position.values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "1a3dc802-06fc-47b4-b596-3f5ddee82e24",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: left;\">\n",
       "      <th></th>\n",
       "      <th>Variant ID</th>\n",
       "      <th>rsID</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Consequence</th>\n",
       "      <th>Annotation</th>\n",
       "      <th>Flags</th>\n",
       "      <th>Allele Count</th>\n",
       "      <th>Allele Number</th>\n",
       "      <th>Allele Frequency</th>\n",
       "      <th>Number of Homozygotes</th>\n",
       "      <th>Source</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Location</th>\n",
       "      <th>Reference</th>\n",
       "      <th>Alternative</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>283</th>\n",
       "      <td>6-33577640-G-T</td>\n",
       "      <td>rs375089143</td>\n",
       "      <td>BAK1</td>\n",
       "      <td>c.-31-5C&gt;A</td>\n",
       "      <td>splice_region_va...</td>\n",
       "      <td>[]</td>\n",
       "      <td>1</td>\n",
       "      <td>151382</td>\n",
       "      <td>0.000007</td>\n",
       "      <td>0</td>\n",
       "      <td>Genome</td>\n",
       "      <td>6</td>\n",
       "      <td>33577640</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>284</th>\n",
       "      <td>6-33577640-G-A</td>\n",
       "      <td>rs375089143</td>\n",
       "      <td>BAK1</td>\n",
       "      <td>c.-31-5C&gt;T</td>\n",
       "      <td>splice_region_va...</td>\n",
       "      <td>[]</td>\n",
       "      <td>3</td>\n",
       "      <td>151382</td>\n",
       "      <td>0.000020</td>\n",
       "      <td>0</td>\n",
       "      <td>Genome</td>\n",
       "      <td>6</td>\n",
       "      <td>33577640</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Variant ID      rsID         Gene  Consequence Annotation           Flags  \\\n",
       "283  6-33577640-G-T  rs375089143  BAK1  c.-31-5C>A  splice_region_va...  []     \n",
       "284  6-33577640-G-A  rs375089143  BAK1  c.-31-5C>T  splice_region_va...  []     \n",
       "\n",
       "     Allele Count  Allele Number  Allele Frequency  Number of Homozygotes  \\\n",
       "283  1             151382         0.000007          0                       \n",
       "284  3             151382         0.000020          0                       \n",
       "\n",
       "    Source  Chromosome Location  Reference Alternative  \n",
       "283  Genome  6          33577640  G         T           \n",
       "284  Genome  6          33577640  G         A           "
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mad_df.loc[mad_df.Location.isin(dels)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "2719112b-887d-468d-a658-26072e23fc75",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: left;\">\n",
       "      <th></th>\n",
       "      <th>Variant ID</th>\n",
       "      <th>rsID</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Consequence</th>\n",
       "      <th>Annotation</th>\n",
       "      <th>Flags</th>\n",
       "      <th>Allele Count</th>\n",
       "      <th>Allele Number</th>\n",
       "      <th>Allele Frequency</th>\n",
       "      <th>Number of Homozygotes</th>\n",
       "      <th>Source</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Location</th>\n",
       "      <th>Reference</th>\n",
       "      <th>Alternative</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>125</th>\n",
       "      <td>8-140662662-CAGG-C</td>\n",
       "      <td>rs1188036104</td>\n",
       "      <td>PTK2</td>\n",
       "      <td>p.Pro51del</td>\n",
       "      <td>inframe_deletion</td>\n",
       "      <td>[]</td>\n",
       "      <td>1</td>\n",
       "      <td>152184</td>\n",
       "      <td>0.000007</td>\n",
       "      <td>0</td>\n",
       "      <td>Genome</td>\n",
       "      <td>8</td>\n",
       "      <td>140662662</td>\n",
       "      <td>CAGG</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>131</th>\n",
       "      <td>8-140662690-AGTC...</td>\n",
       "      <td>rs202056008</td>\n",
       "      <td>PTK2</td>\n",
       "      <td>p.Leu41_Thr42del</td>\n",
       "      <td>inframe_deletion</td>\n",
       "      <td>[]</td>\n",
       "      <td>2643</td>\n",
       "      <td>152190</td>\n",
       "      <td>0.017366</td>\n",
       "      <td>45</td>\n",
       "      <td>Genome</td>\n",
       "      <td>8</td>\n",
       "      <td>140662690</td>\n",
       "      <td>AGTCAAT</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>206</th>\n",
       "      <td>8-140664939-AGGA...</td>\n",
       "      <td>rs1213454475</td>\n",
       "      <td>PTK2</td>\n",
       "      <td>p.Pro986_Leu987del</td>\n",
       "      <td>inframe_deletion</td>\n",
       "      <td>[]</td>\n",
       "      <td>3</td>\n",
       "      <td>142670</td>\n",
       "      <td>0.000021</td>\n",
       "      <td>0</td>\n",
       "      <td>Genome</td>\n",
       "      <td>8</td>\n",
       "      <td>140664939</td>\n",
       "      <td>AGGAGGG</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>500</th>\n",
       "      <td>8-140700924-ATCT-A</td>\n",
       "      <td>rs766236776</td>\n",
       "      <td>PTK2</td>\n",
       "      <td>p.Glu821del</td>\n",
       "      <td>inframe_deletion</td>\n",
       "      <td>[]</td>\n",
       "      <td>1</td>\n",
       "      <td>151988</td>\n",
       "      <td>0.000007</td>\n",
       "      <td>0</td>\n",
       "      <td>Genome</td>\n",
       "      <td>8</td>\n",
       "      <td>140700924</td>\n",
       "      <td>ATCT</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Variant ID           rsID          Gene  Consequence          \\\n",
       "125   8-140662662-CAGG-C  rs1188036104  PTK2          p.Pro51del   \n",
       "131  8-140662690-AGTC...   rs202056008  PTK2    p.Leu41_Thr42del   \n",
       "206  8-140664939-AGGA...  rs1213454475  PTK2  p.Pro986_Leu987del   \n",
       "500   8-140700924-ATCT-A   rs766236776  PTK2         p.Glu821del   \n",
       "\n",
       "    Annotation        Flags  Allele Count  Allele Number  Allele Frequency  \\\n",
       "125  inframe_deletion  []       1          152184         0.000007           \n",
       "131  inframe_deletion  []    2643          152190         0.017366           \n",
       "206  inframe_deletion  []       3          142670         0.000021           \n",
       "500  inframe_deletion  []       1          151988         0.000007           \n",
       "\n",
       "     Number of Homozygotes Source  Chromosome Location   Reference Alternative  \n",
       "125   0                     Genome  8          140662662     CAGG   C           \n",
       "131  45                     Genome  8          140662690  AGTCAAT   A           \n",
       "206   0                     Genome  8          140664939  AGGAGGG   A           \n",
       "500   0                     Genome  8          140700924     ATCT   A           "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rs = GeneSearch(3, 'PTK2')\n",
    "mad_df, clinical_df = rs.get_data()\n",
    "mad_df.loc[mad_df.Annotation == \"inframe_deletion\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "4ab71840-7d36-494f-b950-4f21671168c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "reg = 42694920\n",
    "df = Search_region('hg38', 5, reg-50, reg+50)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "f32de09a-bfb7-4d93-9f61-3a85295fe08a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: left;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Position</th>\n",
       "      <th>Reference</th>\n",
       "      <th>Alternative</th>\n",
       "      <th>Annotation</th>\n",
       "      <th>Gene</th>\n",
       "      <th>GATK Filter</th>\n",
       "      <th>Number of Homozygotes</th>\n",
       "      <th>Allele Number</th>\n",
       "      <th>Allele Count</th>\n",
       "      <th>Allele Frequency</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>5</td>\n",
       "      <td>42694908</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>intronic</td>\n",
       "      <td>GHR</td>\n",
       "      <td>PASS</td>\n",
       "      <td>0</td>\n",
       "      <td>2342</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000427</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>5</td>\n",
       "      <td>42694923</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>synonymous SNV</td>\n",
       "      <td>GHR</td>\n",
       "      <td>PASS</td>\n",
       "      <td>0</td>\n",
       "      <td>2342</td>\n",
       "      <td>2</td>\n",
       "      <td>0.000854</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Chromosome Position  Reference Alternative Annotation      Gene GATK Filter  \\\n",
       "0  5          42694908  T         C                 intronic  GHR  PASS         \n",
       "1  5          42694923  T         A           synonymous SNV  GHR  PASS         \n",
       "\n",
       "  Number of Homozygotes Allele Number Allele Count Allele Frequency  \n",
       "0  0                     2342          1            0.000427         \n",
       "1  0                     2342          2            0.000854         "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "aadf1dac-93c1-4b64-856d-5829b4dfa850",
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "import sys\n",
    "\n",
    "def genome_ref_info(chromosome,version,position):\n",
    "       \n",
    "    if version in [\"hg38\",\"GRCh38\"]:\n",
    "        server = \"https://rest.ensembl.org/sequence/region/human/%s:%d..%d:1?\" %(chromosome,position,position)\n",
    "    if version in [\"hg19\", \"GRCh37\"]:\n",
    "        server=  \"https://grch37.rest.ensembl.org/sequence/region/human/%s:%d..%d:1?\" %(chromosome,position,position)\n",
    "       \n",
    "    request = requests.get(server, headers={ \"Content-Type\" : \"text/plain\"})\n",
    "    if not request.ok:\n",
    "        request.raise_for_status()\n",
    "        sys.exit()\n",
    "    return(request.text)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "71e7cacd-4850-492e-b6e4-43ebd14cb8ca",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'A'"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genome_ref_info(1, \"hg38\", 32332725-1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "28963731-ea03-43af-9644-59eb119a7184",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: left;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Position</th>\n",
       "      <th>Reference</th>\n",
       "      <th>Alternative</th>\n",
       "      <th>Annotation</th>\n",
       "      <th>Gene</th>\n",
       "      <th>GATK Filter</th>\n",
       "      <th>Number of Homozygotes</th>\n",
       "      <th>Allele Number</th>\n",
       "      <th>Allele Count</th>\n",
       "      <th>Allele Frequency</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>5</td>\n",
       "      <td>42694908</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>intronic</td>\n",
       "      <td>GHR</td>\n",
       "      <td>PASS</td>\n",
       "      <td>0</td>\n",
       "      <td>2342</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000427</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>5</td>\n",
       "      <td>42694923</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>synonymous SNV</td>\n",
       "      <td>GHR</td>\n",
       "      <td>PASS</td>\n",
       "      <td>0</td>\n",
       "      <td>2342</td>\n",
       "      <td>2</td>\n",
       "      <td>0.000854</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Chromosome Position  Reference Alternative Annotation      Gene GATK Filter  \\\n",
       "0  5          42694908  T         C                 intronic  GHR  PASS         \n",
       "1  5          42694923  T         A           synonymous SNV  GHR  PASS         \n",
       "\n",
       "  Number of Homozygotes Allele Number Allele Count Allele Frequency  \n",
       "0  0                     2342          1            0.000427         \n",
       "1  0                     2342          2            0.000854         "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "604d8183-4efc-43f7-ac0a-30620e78d33f",
   "metadata": {},
   "outputs": [],
   "source": [
    "d = {\n",
    "    \"Variant ID\": [\"aaa\", \"bb\"],\n",
    "    \"Position\": [1111, 22222],\n",
    "    \"Reference\": [\"T\", \"A\"]\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "bb5c3dbf-aba8-47aa-98e8-1144a2cc6772",
   "metadata": {},
   "outputs": [],
   "source": [
    "df[\"Position\"] = [1111, 22222]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "63ba98ce-8b89-447c-a9e8-ccfb77b51667",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: left;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Position</th>\n",
       "      <th>Reference</th>\n",
       "      <th>Alternative</th>\n",
       "      <th>Annotation</th>\n",
       "      <th>Gene</th>\n",
       "      <th>GATK Filter</th>\n",
       "      <th>Number of Homozygotes</th>\n",
       "      <th>Allele Number</th>\n",
       "      <th>Allele Count</th>\n",
       "      <th>Allele Frequency</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>5</td>\n",
       "      <td>1111</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>intronic</td>\n",
       "      <td>GHR</td>\n",
       "      <td>PASS</td>\n",
       "      <td>0</td>\n",
       "      <td>2342</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000427</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>5</td>\n",
       "      <td>22222</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>synonymous SNV</td>\n",
       "      <td>GHR</td>\n",
       "      <td>PASS</td>\n",
       "      <td>0</td>\n",
       "      <td>2342</td>\n",
       "      <td>2</td>\n",
       "      <td>0.000854</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Chromosome  Position Reference Alternative Annotation      Gene GATK Filter  \\\n",
       "0  5           1111     T         C                 intronic  GHR  PASS         \n",
       "1  5          22222     T         A           synonymous SNV  GHR  PASS         \n",
       "\n",
       "  Number of Homozygotes Allele Number Allele Count Allele Frequency  \n",
       "0  0                     2342          1            0.000427         \n",
       "1  0                     2342          2            0.000854         "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "85e19b9d-8824-41f1-9862-bc4d113d1213",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
